juju/juju
# fix: allow fast actions to complete quickly

We put it a reasonable backoff of 2.0s when polling for actions being completed, except many actions actually complete very quickly.

On my hardware, it takes about 40ms to enqueue an action (including Login, etc), and then an action like `echo hello` takes about another 40ms to complete and be recorded.
Which meant that while the action itself actually was completing in about 100ms wall clock time, we were always waiting 2080ms before telling the user about it.

This should make all those times when people want to run `juju exec --unit foo/0 -- relation-get` much more interactive and pleasant to use.

On my hardware, this drops the `juju exec` time to 100ms.

fixes: LP#2078357

## Checklist




## QA steps

There are 2 key QA steps:

```sh
time juju exec -m controller --unit controller/0 -- echo hello
```

Without this patch, those take just slightly longer than 2.0s on my machine (my machine takes roughly 2.080s.)
With this patch, it can complete in around 100ms on my machine.

Try exploring around the juju model and see how it feels:
```sh
juju deploy juju-qa-dummy-sink
juju deploy juju-qa-dummy-source
juju relate dummy-sink dummy-source
juju exec --unit dummy-sink/0 relation-ids source
juju exec --unit dummy-sink/0 relation-list -r source:0 
juju exec --unit dummy-sink/0 relation-get -r source:0 - dummy-source/0
```

This also has an impact when running builtin actions:
```sh
juju deploy ubuntu-lite
time juju run ubuntu-lite/0 load
```

On my machine Juju 3.5.3 takes:
```
real    0m2.119s
```

while this patch changes it to:
```
real    0m0.469s
```

If you want to see what requests are actually being made, you can do:
```sh
juju --debug  --logging-config "juju.rpc=TRACE" exec -m controller --unit controller/0 -- echo hello
```
That will print out each of the query and responses that we do/get. I also did put the retry logic into a debug message, because it seemed roughly helpful to be able to do:
```sh
$ time ~/go/bin/juju --debug exec --unit "ubuntu-lite/0" -- echo hello
...
14:37:43 DEBUG juju.cmd.juju.action common.go:462 after 947.584µs action was still pending, will wait 20ms more before next check
14:37:43 DEBUG juju.cmd.juju.action common.go:462 after 21.14602ms action was still running, will wait 30ms more before next check
```

## Documentation changes

Shouldn't require any documentation changes.

## Links

**Launchpad bug:** https://bugs.launchpad.net/juju/+bug/2078357
